bson/src/main/org/bson/codecs/configuration/mapper/ClassModel.java

```diff
@@ -0,0 +1,78 @@
+package org.bson.codecs.configuration.mapper;
+
+import com.fasterxml.classmate.MemberResolver;
+import com.fasterxml.classmate.ResolvedType;
+import com.fasterxml.classmate.ResolvedTypeWithMembers;
+import com.fasterxml.classmate.TypeResolver;
+import com.fasterxml.classmate.members.ResolvedField;
+import org.bson.BsonReader;
+import org.bson.BsonWriter;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class ClassModel extends MappedType {
+    private final Map<String, FieldModel> fields = new TreeMap<String, FieldModel>();
+    private final CodecRegistry registry;
+    private final TypeResolver resolver;
+    
+    public ClassModel(final CodecRegistry registry, final TypeResolver resolver, final Class<?> aClass) {
+        super(aClass);
+        this.registry = registry;
+        this.resolver = resolver;
+        map();
+    }
+
+    public <T> T decode(final BsonReader reader, final DecoderContext decoderContext) {
```

Comment:
delegating down to where the mapping information is.  I'm not entirely convinced the two classes shouldn't be merged.  ClassModelCodec feels like a thin, nigh-useless veneer on ClassModel.